allennlp/data/tokenizers/pretrained_transformer_tokenizer.py

```diff
@@ -177,139 +276,153 @@ def _tokenize(self, sentence_1: str, sentence_2: str = None):
                     continue
                 allowed_skipped_whitespace = min_allowed_skipped_whitespace
 
-                tokens[token_index].idx = token_start_index
+                token_offsets[token_index] = (
+                    token_start_index,
+                    token_start_index + len(token_text),
+                )
                 text_index = token_start_index + len(token_text)
                 token_index += 1
 
-        return tokens
+        tokens = []
+        for token_id, token_type_id, offsets in zip(token_ids, token_type_ids, token_offsets):
+            if offsets is None or offsets[0] >= offsets[1]:
+                token_str = self.tokenizer.convert_ids_to_tokens(
+                    token_id, skip_special_tokens=False
+                )
+                start = None
+            else:
+                start, end = offsets
+                token_str = text[start:end]
 
-    def tokenize_sentence_pair(self, sentence_1: str, sentence_2: str) -> List[Token]:
-        """
-        This methods properly handles a pair of sentences.
-        """
-        return self._tokenize(sentence_1, sentence_2)
+            tokens.append(Token(text=token_str, text_id=token_id, type_id=token_type_id, idx=start))
```

Comment:
Sorry if I've missed or forgotten an explanation you already gave, but I don't really follow why this needs to be `token_str`, instead of just doing a reverse lookup on `token_id`, as (I believe) we had before.  What do we gain from this?  What do we lose?  (I gave one of the latter in other comments on this PR.)

I'm _pretty sure_ that nothing else in the indexing pipeline will look at `token.text`, because the pretrained transformer indexer can just look at `text_id` and `type_id`, and ignore `token.text`.  So what's the point of this change?  Is it to make offsets easier?  So you can do `idx + len(text)` in order to get character offsets?  If that's the problem that's being solved here, seems easier to just change `idx` to `start_index` and `end_index`, and we can keep the semantics of `token.text` the same.

And, sorry that I didn't realize this issue earlier.

Reply:
The original purpose was so that the length is correct. But I also like the intuition that you can concatenate tokens and get real text back.

On the other hand, I like the intuition that there is a 1:1 mapping between token string and token id. We can't have both.

Reply:
Well, we can have both by just adding another (optional) field to `Token`, like `display_text`, or something.  And beyond intuition is the fact that we have parts of our code that rely on the 1:1 mapping, and it's not clear to me how I could modify that logic to handle the case where we lose the 1:1 mapping.  It's pretty easy to think of how to modify whatever code depends on the length being right, if there is any.

Reply:
The whole `Token` class doesn't make sense for transformers. What's `pos_` or `lemma_` on a word piece?

After 1.0 ships we need to do some real clean-up of this API. I'm beginning to formulate ideas for 2.0.

Meanwhile, for 1.0, I'll add an `idx_end` field to track this.

Reply:
I changed this back so that `token.text` is the word piece, and I added `token.idx_end`, which is one past the last character of the token.